docs/ExtensibleMaps.md

```diff
@@ -0,0 +1,80 @@
+
+## Verfication
+- For offline verification, eBPF store will be used to provide map information to the verfier.
+- For online verification (JIT / interpret), map type to program type conversion will be done using eBPF store data. Then ebpfsvc will query ebpfcore to get program information which will also provide map information.
+
+## Map lifecycle
+Even though the extensible map will be created by and reside in the extension, ebpfcore will also create a corresponding map entry, as it does for the global maps. The difference being, in case of extensible maps, the map CRUD APIs will be supplied by the extension, and map entry in ebpfcore will contain these function pointers provided by the extension.
+
+Map lifetime will also be maintained by eBPFcore, and it will invoke extension's map delete API when the map needs to be finally deleted.
+Similarly, map pinning will also be handled by eBPFcore as that impacts map lifetime.
+
+Another thing to note is that once an extensible map is created, the corresponding extension cannot be allowed to unload / restart, as that will delete the map and its entries. This will be a limitation / restriction for the extension that is implementing extensible maps, and may impact their servicing flow.
+
+## Map creation
+Assuming option 1 for `Map ID partitioning`, below is the expected flow for map creation.
+
+### Explicit map creation
+- App uses the existing map create APIs, and internally ebpfapi tries to find the corresponding program type from the eBPF store.
+- Once it finds the program info provider, it makes a (new) ioctl call to create the extensible map, and also pass the program type.
+- eBPFcore will first attach (NMR) to this provider, and check if the actual provider supports this map type. If yes, proceed to create map in the extension.
+
+Implicit map creation flow will also be similar. ebpf runtime will have similar flow for map creation, automatic map pinning, and map reuse.
+
+## Map CRUD APIs
+
+### Usermode CRUD APIs
+Assuming option 1 for `Map ID partitioning`, all existing APIs should be applicable for extensible maps too.
+
+### Map helper functions
+The existing map helper functions implemented by ebpfcore will be used by the BPF programs for extensible maps also. For extensible maps, ebpfcore will redirect the calls to the extension.
+
+## Exposing RCU semantics to extensions
+For extensions to implement maps, they will need RCU support, and eBPF needs to expose RCU / epoch logic to extensions.
+
+There are two options for this:
+
+**Export RCU as lib**
+- This will simplify logic in eBPFCore
+- Extensions will have their own RCU "runtime"
+- This approach will require recompilation and release from extension if there is a bug in RCU lib.
+
+**Export RCU APIs via NMR interface**
+- Probably adds more complexity to ebpfcore.
+- Does not require new release from extensions wheenver there is a bugfix in RCU logic.
+
+**Proposal**
+Proposal here is to export RCU as lib.
```

Comment:
I am not sure having two copies of the RCU logic running at the same time makes a lot of sense.